### PR TITLE
feat(HACBS-1747): update binding if one already exists

### PR DESCRIPTION
**Pull request by johnbieren**
When the release service processes multiple snapshots for the same application and environment, it should create one binding and then update it for each successive snapshot. This commit changes the behavior from creating a new binding each time to the desired behavior.

Signed-off-by: Johnny Bieren <jbieren@redhat.com>

### CODE REVIEW

**Comment by scoheb** [comment]
Is this dependent on https://github.com/redhat-appstudio/release-service/pull/166 ?

**Comment by johnbieren** [comment]
> Is this dependent on #166 ?

No, it isn't dependent on #166. In fact, I think with this, we don't even need #166 anymore (I may be wrong though). #166 is about 
release 1 with application 1 and snapshot 1 creates binding 1
release 2 with application 1 and snapshot 2 needs to create binding 2 so we can't just rely on application

However, now, release 2 with application 1 and snapshot 2 should update binding 1, so we do actually want to retrieve binding 1 in the get (so again ignore the snapshot when retrieving)

File: controllers/release/release_adapter.go [hunk 1]
```diff
@@ -225,34 +225,23 @@ func (a *Adapter) EnsureSnapshotEnvironmentBindingExists() (reconciler.Operation
...
-		binding, err := a.createSnapshotEnvironmentBinding(releasePlanAdmission)
-		if err != nil {
-			return reconciler.RequeueWithError(err)
-		}
-
-		a.logger.Info("Created SnapshotEnvironmentBinding",
-			"SnapshotEnvironmentBinding.Name", binding.Name, "SnapshotEnvironmentBinding.Namespace", binding.Namespace)
+	binding, err := a.createOrUpdateSnapshotEnvironmentBinding(releasePlanAdmission)
+	if err != nil {
+		return reconciler.RequeueWithError(err)
```
[the remark this answers is not in the document]
Comment (johnbieren) [reply controllers/release/release_adapter.go:235]: Yeah, I see them too. I am not sure how codecov wants us to fix it. I already have tests in the context of `createOrUpdateSnapshotEnvironmentBinding` for the error scenarios, and this line is just about that function failing, so not sure

[the remark this answers is not in the document]
Comment (happybhati) [reply controllers/release/release_adapter.go:235]: Exactly, that's what I was wondering about. Thanks for the Info.

[the remark this answers is not in the document]
Comment (mmalina) [reply controllers/release/release_adapter.go:235]: I'm not saying we really need it. But to satisfy this we would need a new test for `EnsureSnapshotEnvironmentBindingExists` which would mock `createOrUpdateSnapshotEnvironmentBinding` to return an error and then we would check that the function returned `reconciler.RequeueWithError()` with the error, right?

[the remark this answers is not in the document]
Comment (davidmogar) [reply controllers/release/release_adapter.go:235]: There is stuff that, IMHO, are not worth it to cover. At least at the moment. This includes this line.

File: controllers/release/release_adapter_test.go [hunk 2]
```diff
@@ -602,6 +615,26 @@ var _ = Describe("Release Adapter", Ordered, func() {
 		})
 
 		It("tracks the binding if one is found", func() {
+			adapter.ctx = loader.GetMockedContext(ctx, []loader.MockData{
+				{
+					ContextKey: loader.SnapshotEnvironmentBindingContextKey,
+					Resource: &applicationapiv1alpha1.SnapshotEnvironmentBinding{
+						ObjectMeta: metav1.ObjectMeta{
```
Comment (davidmogar) [inline controllers/release/release_adapter_test.go:622]: I don't think you need the ObjectMeta at all. The binding is loaded from the Release status.

File: controllers/release/release_adapter.go [hunk 3]
```diff
@@ -345,6 +356,12 @@ func (a *Adapter) registerGitOpsDeploymentStatus(binding *applicationapiv1alpha1
 		return nil
 	}
 
+	// Do nothing if the release does not own the binding
+	if binding.GetAnnotations()[libhandler.TypeAnnotation] != a.release.GetObjectKind().GroupVersionKind().GroupKind().String() ||
```
Comment (davidmogar) [inline controllers/release/release_adapter.go:360]: This check should be in the operation itself and call `registerGitOpsDeploymentStatus` with a valid binding only. First time I went over the operation made me wonder if you were taking into account the ownership.

**Review by davidmogar — CHANGES_REQUESTED** [review]

File: controllers/release/release_adapter.go (hunk 3 above)
Comment (johnbieren) [reply controllers/release/release_adapter.go]: fixed

**Review by davidmogar — APPROVED** [review]

**Pull request merged**